### ISSUE TITLE
FYI: comments on reproducible builds on Ubuntu 18.04 bionic and 18.10 cosmic

### ISSUE THREAD

**Issue by riyad** [Eloston/ungoogled-chromium#632]
I've used [diffoscope](https://diffoscope.org/) to compare two of my cosmic builds and @Eloston's bionic build against my own and I must say ungoogled-chromium seems reproducible.

### Cosmic
The differences are:
- all deb files have the same content/hash :heavy_check_mark: 
- all files have different file modification times. This is expected. As it's only metadata this doesn't change the file contents. :heavy_check_mark: 
- `foo.buildinfo` records the build time. This is **the only difference** between both files' content. :neutral_face: 
- which sadly `foo.changes` records as it saves the hash of every file. Hence this file's content differs in the recorded file hash of the `foo.buildinfo` file. :disappointed:

[cosmic.diffoscope.txt](https://github.com/Eloston/ungoogled-chromium/files/2683840/cosmic.diffoscope.txt)

### Bionic
- all deb files have the same content/hash :heavy_check_mark: although @Eloston's debs seem to be named slightly differently (`.bionic` vs `~bionic`).
- all files have different file modification times and group permissions. This is expected (except for permissions). As it's only metadata this doesn't change the file contents. :heavy_check_mark:
- Since `foo.changes`  records the hash of every file. This file's content differs in the recorded file hash of the `foo.buildinfo` file. :disappointed:
- Although the produced deb files are the same the `foo.buildinfo` seems to have quite a few differences.
  - time of build
  - several differences in the `Installed-Build-Depends:` section. I'm not sure how those can differ. I just installed all packages listed by `dpkg-checkbuilddeps`. This still didn't seem to impact the resulting binaries though. o.O

[bionic.diffoscope.txt](https://github.com/Eloston/ungoogled-chromium/files/2683861/bionic.diffoscope.txt)

### Builds
All builds were done in fresh LXD/LXC containers with the respective distro image. Bionic builds additionally used upstream LLVM. I can provide [Ansible](https://www.ansible.com/) roles if anybody is interested.

**Comment by Eloston** [comment]
Thanks for testing and writing up your results. I am new to the subject of reproducible builds, so these results are interesting to me.

In regards to the differences, you mentioned in both bionic and cosmic how it is just the `buildinfo` and `changes` files that vary. These are just plain-text metadata files, so I don't think it matters if they are identical. However, I still think we should consult Debian's documentation on reproducible builds for their practices here.

That being said, what would you like us to do with this information? Do you think ungoogled-chromium should start applying ideas of reproducible builds?

**Comment by riyad** [comment]
Me neither, but the warning at the top of the ungoogled-chromium downloads page sparked my interest. Once I had the build process somewhat automated I just tried to poke around diffing two of my builds out of curiosity. I was surprised by the outcome to be honest. I was even more surprised when comparing my build against yours. :slightly_smiling_face: 

I had watched a few talks about the reproducible builds effort and was aware that there are a myriad of possibilities for differences especially in a project the size of chromium. And a little research shows that [chromium itself isn't reproducible yet](https://tests.reproducible-builds.org/debian/rb-pkg/unstable/amd64/chromium.html) ... it seems because of build ids and some debug info. Whatever the source, it seems it's patched out of ungoogled-chromium. :smile: 

I'm still not sure what the consequence should be. We should be aware that builds on other platforms might not be reproducible ... although I'm fairly confident they are since AFAIK Windows and macOS builds also use the LLVM toolchain (instead of "platform specific" ones).

One idea would be to ask previous/new contributors of binaries to try to verify on their platform/distro.
Another would be to invert the assumption on the download page to e.g. "(preliminary) tests have shown builds should be/can be expected to be reproducible" and ask users for ideas and feedback.
Maybe gathering some more data and/or making this info public will spark some idea. :man_shrugging:

I was just the lucky finder ... I don't know what to do with it. :joy:
